dbms/src/Flash/Mpp/MPPTunnelSet.cpp

```diff
@@ -63,83 +63,200 @@ void MPPTunnelSetBase<Tunnel>::write(tipb::SelectResponse & response)
+    auto remote_tracked_packet_bytes = remote_tracked_packet ? remote_tracked_packet->getPacket().ByteSizeLong() : 0;
+    checkPacketSize(ori_tracked_packet_bytes);
+    checkPacketSize(remote_tracked_packet_bytes);
+
+    // TODO avoid copy packet for broadcast.
 
-    switch (method)
+    if (!remote_tracked_packet)
     {
-    case CompressionMethod::NONE:
+        assert(ori_tracked_packet);
+
+        remote_tracked_packet_bytes = ori_tracked_packet_bytes;
+
+        for (size_t i = 1; i < tunnels.size(); ++i)
+            tunnels[i]->write(ori_tracked_packet->copy());
+        tunnels[0]->write(std::move(ori_tracked_packet));
+    }
+    else
     {
-        if (is_local)
-        {
-            GET_METRIC(tiflash_exchange_data_bytes, type_hash_none_compression_local).Increment(sz);
-        }
-        else
+        for (size_t i = 0, local_cnt = 0, remote_cnt = 0; i < tunnels.size(); ++i)
         {
-            GET_METRIC(tiflash_exchange_data_bytes, type_hash_none_compression_remote).Increment(sz);
+            if (IsLocalTunnel(tunnels[i]))
+            {
+                local_cnt++;
+                if (local_cnt == local_tunnel_cnt)
+                    tunnels[i]->write(std::move(ori_tracked_packet));
+                else
+                    tunnels[i]->write(ori_tracked_packet->copy()); // NOLINT
+            }
+            else
+            {
+                remote_cnt++;
+                if (remote_cnt == remote_tunnel_cnt)
+                    tunnels[i]->write(std::move(remote_tracked_packet));
+                else
+                    tunnels[i]->write(remote_tracked_packet->copy()); // NOLINT
+            }
         }
-        break;
     }
-    case CompressionMethod::LZ4:
+
+    if (is_broadcast)
     {
-        GET_METRIC(tiflash_exchange_data_bytes, type_hash_lz4_compression).Increment(sz);
-        break;
+        UPDATE_EXCHANGE_MATRIC(broadcast, CompressionMethod::NONE, local_tunnel_cnt * ori_tracked_packet_bytes, local_tunnel_cnt * ori_tracked_packet_bytes, true);
+        UPDATE_EXCHANGE_MATRIC(broadcast, compression_method, remote_tunnel_cnt * ori_tracked_packet_bytes, remote_tunnel_cnt * remote_tracked_packet_bytes, false);
     }
-    case CompressionMethod::ZSTD:
+    else
     {
-        GET_METRIC(tiflash_exchange_data_bytes, type_hash_zstd_compression).Increment(sz);
-        break;
-    }
-    default:
-        break;
+        UPDATE_EXCHANGE_MATRIC(passthrough, CompressionMethod::NONE, local_tunnel_cnt * ori_tracked_packet_bytes, local_tunnel_cnt * ori_tracked_packet_bytes, true);
+        UPDATE_EXCHANGE_MATRIC(passthrough, compression_method, remote_tunnel_cnt * ori_tracked_packet_bytes, remote_tunnel_cnt * remote_tracked_packet_bytes, false);
     }
 }
 
 template <typename Tunnel>
-void MPPTunnelSetBase<Tunnel>::broadcastOrPassThroughWrite(Blocks & blocks)
+void MPPTunnelSetBase<Tunnel>::broadcastOrPassThroughWrite(Blocks & blocks, bool is_broadcast)
```

Comment:
will it be more clear if we distinguish broadcast and passthrough?

Reply:
Fixed.